### PR TITLE
Phase 6: Learnable Asinh Scale — Adaptive Pressure Compression

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current asinh pressure transform uses a fixed `asinh_scale=0.75`, chosen by grid search (PR #2054). The compression function `asinh(p * s)` has different effects at different pressure magnitudes: too large over-compresses peaks, too small under-compresses. Making `asinh_scale` a learnable `nn.Parameter` allows gradient descent to find the optimal compression jointly with model weights.

**Why this might help:** For tandem foils (p_tan=29.1, our worst metric), pressure distributions differ significantly between inner and outer channels. A learned scale can adapt per-run to the actual distribution. Extension: asymmetric asinh with separate scales for positive/negative pressure.

### Code changes (~5 lines)

1. In `Transolver.__init__` (or wherever the model is initialized), replace the fixed scale with a learnable parameter:
```python
# Replace fixed scale usage with:
self.asinh_scale = nn.Parameter(torch.tensor(cfg.asinh_scale))  # init to 0.75
```

2. In the pressure preprocessing (where `torch.asinh(p * cfg.asinh_scale)` is applied):
```python
# Use softplus to keep scale positive
scale = F.softplus(self.asinh_scale)
p_transformed = torch.asinh(p * scale)
```

3. In the inverse transform (prediction postprocessing):
```python
scale = F.softplus(self.asinh_scale)
p_pred = torch.sinh(p_transformed) / scale
```

4. Log the learned scale value to W&B each epoch: `wandb.log({"asinh_scale": F.softplus(self.asinh_scale).item()})`

### Experiment matrix (8 GPUs)

| GPU | Config | Seed | Notes |
|-----|--------|------|-------|
| 0 | Baseline (fixed s=0.75) | 42 | Control |
| 1 | Baseline (fixed s=0.75) | 43 | Control |
| 2 | Learnable scale (init 0.75) | 42 | Main test |
| 3 | Learnable scale (init 0.75) | 43 | Main test |
| 4 | Learnable scale (init 0.75) | 44 | More seeds |
| 5 | Learnable scale (init 0.75) | 45 | More seeds |
| 6 | Asymmetric (2 scales: pos/neg init 0.75) | 42 | Extension |
| 7 | Asymmetric (2 scales: pos/neg init 0.75) | 43 | Extension |

For the **asymmetric variant** (GPUs 6-7):
```python
self.asinh_scale_pos = nn.Parameter(torch.tensor(0.75))
self.asinh_scale_neg = nn.Parameter(torch.tensor(0.75))

# In preprocessing:
scale_pos = F.softplus(self.asinh_scale_pos)
scale_neg = F.softplus(self.asinh_scale_neg)
scale = torch.where(p >= 0, scale_pos, scale_neg)
p_transformed = torch.asinh(p * scale)
```

Use `--wandb_group phase6/learnable-asinh` for all runs.
Use a new flag like `--learnable_asinh` (and `--asymmetric_asinh` for GPUs 6-7).

**Base command:**
```bash
python train.py --agent thorfinn --wandb_group phase6/learnable-asinh \
  --wandb_name "thorfinn/learnable-asinh-s${seed}" \
  --learnable_asinh \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed ${seed}
```

### Important notes
- DO NOT modify MAX_EPOCHS (500) or MAX_TIMEOUT (180.0) in train.py
- The softplus constraint is important — prevents scale from going negative
- Watch the W&B log for the learned scale value — does it stay near 0.75 or drift?
- If scale drifts significantly, that's evidence the fixed value was suboptimal

## Baseline

Current best (8-seed ensemble, PR #2080):
- p_in: **12.2** | p_oodc: **6.7** | p_tan: **29.1** | p_re: **5.8**

Single-model baseline (8-seed mean):
- p_in: **13.03** | p_oodc: **7.83** | p_tan: **30.29** | p_re: **6.45**

## Reporting

Post results with:
- Individual run metrics for all 8 configs
- Mean ± std for learnable vs baseline
- Final learned scale value(s) for each run
- Whether the asymmetric variant learns different pos/neg scales
- W&B run IDs

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Summary: Learnable asinh scale **fails** — scale collapses to near-zero, degrading all metrics.

### Individual Run Metrics

| GPU | Config | Seed | p_in | p_oodc | p_tan | p_re | val_loss | Best Epoch | Final Scale |
|-----|--------|------|------|--------|-------|------|----------|------------|-------------|
| 0 | **Baseline** (fixed s=0.75) | 42 | **12.8** | **7.7** | **30.1** | **6.4** | 0.386 | 156 | 0.75 (fixed) |
| 1 | **Baseline** (fixed s=0.75) | 43 | **13.3** | **7.9** | **29.8** | **6.6** | 0.387 | 157 | 0.75 (fixed) |
| 2 | Learnable (init 0.75) | 42 | 25.0 | 21.0 | 35.6 | 16.9 | 0.081 | 153 | **0.0028** |
| 3 | Learnable (init 0.75) | 43 | 24.7 | 19.5 | 34.2 | 16.4 | 0.081 | 151 | **0.0036** |
| 4 | Learnable (init 0.75) | 44 | 27.1 | 23.3 | 36.7 | 19.0 | 0.082 | 152 | **0.0033** |
| 5 | Learnable (init 0.75) | 45 | 27.5 | 22.0 | 35.6 | 18.6 | 0.081 | 151 | **0.0037** |
| 6 | Asymmetric (init 0.75) | 42 | 20.0 | 15.1 | 33.1 | 12.4 | 0.082 | 153 | pos=**0.0063**, neg=**0.0047** |
| 7 | Asymmetric (init 0.75) | 43 | 19.3 | 14.5 | 31.6 | 11.8 | 0.084 | 153 | pos=**0.0097**, neg=**0.0069** |

### Group Comparison

| Metric | Baseline (2 seeds) | Learnable (4 seeds) | Change | Asymmetric (2 seeds) | Change |
|--------|-------------------|--------------------|---------|--------------------|---------|
| p_in | **13.1** | 26.1 | **+100%** | 19.6 | **+51%** |
| p_oodc | **7.8** | 21.4 | **+175%** | 14.8 | **+90%** |
| p_tan | **30.0** | 35.5 | **+19%** | 32.4 | **+8%** |
| p_re | **6.5** | 17.7 | **+173%** | 12.1 | **+86%** |

### Root Cause: Scale Collapse

All learnable runs exhibit **scale collapse**: the asinh_scale parameter dropped from 0.75 → ~0.003 within the first ~40 epochs, and continued to near-zero. The W&B histograms show a monotonic decline (█▇▅▃▂▁).

**Why it collapses:** When scale → 0, `asinh(p * 0) → 0` for all pressures. The loss in the normalized target space becomes artificially small because all targets cluster near zero, which is easy to predict. The gradient from the loss function incentivizes the model to compress the target space rather than learn the true pressure distribution. The softplus constraint prevents exact zero but doesn't prevent collapse to very small values (0.003).

**Why val/loss looks lower but metrics are worse:** The val/loss (0.081 vs 0.386) is misleading — it's computed in the compressed target space. When predictions are inverse-transformed back to physical units, the metrics are much worse because the model learned to predict in a nearly-zero target space.

**Asymmetric variant fares slightly better** (p_in=19.6 vs 26.1) — the two separate scales provide a bit more resistance to full collapse, and interestingly the positive scale collapses slower than the negative scale.

### W&B Run IDs

| GPU | Config | Run ID |
|-----|--------|--------|
| 0 | Baseline s42 | zwyyvhsb |
| 1 | Baseline s43 | npsbhjhg |
| 2 | Learnable s42 | 6noqttqf |
| 3 | Learnable s43 | 967i5akk |
| 4 | Learnable s44 | 16jzy7gf |
| 5 | Learnable s45 | crruke85 |
| 6 | Asymmetric s42 | wbrgp3oh |
| 7 | Asymmetric s43 | yylvcbff |

### Command Used

```bash
# Baseline controls (GPUs 0-1)
CUDA_VISIBLE_DEVICES=$gpu python train.py --agent thorfinn \
  --wandb_group phase6/learnable-asinh --wandb_name "thorfinn/baseline-s${seed}" \
  --asinh_pressure --asinh_scale 0.75 [all baseline flags] --seed ${seed}

# Learnable (GPUs 2-5)
CUDA_VISIBLE_DEVICES=$gpu python train.py --agent thorfinn \
  --wandb_group phase6/learnable-asinh --wandb_name "thorfinn/learnable-s${seed}" \
  --learnable_asinh --asinh_pressure --asinh_scale 0.75 [all baseline flags] --seed ${seed}

# Asymmetric (GPUs 6-7)
CUDA_VISIBLE_DEVICES=$gpu python train.py --agent thorfinn \
  --wandb_group phase6/learnable-asinh --wandb_name "thorfinn/asymmetric-s${seed}" \
  --learnable_asinh --asymmetric_asinh --asinh_pressure --asinh_scale 0.75 [all baseline flags] --seed ${seed}
```

Peak memory: ~38 GB per GPU (same as baseline).

### What Happened

The learnable asinh scale hypothesis **failed** due to scale collapse. Making the asinh scale a free learnable parameter creates a trivial optimization shortcut: the model minimizes loss by compressing the target space to near-zero, rather than by improving predictions. This is analogous to a normalizing flow with a collapsing Jacobian.

The normalization stats mismatch (pre-computed with fixed 0.75, but training with drifting scale) likely exacerbated the issue, as the z-score normalization became meaningless once the scale drifted far from its initialization.

### Suggested Follow-ups

If this direction is still worth pursuing, potential fixes for scale collapse:
1. **Lower bound clamp**: `F.softplus(raw).clamp(min=0.3, max=1.5)` — hard constraint on the range
2. **Regularization**: Add `lambda * (1/scale)` penalty to the loss to resist collapse
3. **Warmup then freeze**: Let scale learn for first N epochs, then freeze it for the rest of training
4. **Discrete grid**: Instead of continuous optimization, do a coarse grid search over s ∈ {0.5, 0.75, 1.0, 1.25} with a few epochs each, then commit
5. **Per-channel learned scales (not per-sign)**: Learn separate scales for surface vs. volume pressure

However, given the magnitude of the failure (~2x worse on all metrics), the fundamental approach of making the target transform learnable appears problematic without strong constraints. The fixed s=0.75 from grid search is likely near-optimal.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2096 because ADVISOR: **Closing — dead end (scale collapse).**

Excellent analysis, thorfinn. The root cause is clear: making the asinh scale a free learnable parameter creates a trivial optimization shortcut — the model minimizes loss by compressing the target space to near-zero rather than learning pressure predictions. All metrics regressed 50-175%.

**Key takeaways for the research programme:**
- The fixed asinh scale s=0.75 from grid search is near-optimal and should remain fixed
- Target-space transforms must NOT be jointly optimized with model weights without strong constraints (the gradient incentive to collapse dominates)
- The val/loss vs physical-metric disconnect (0.081 vs 0.386 but 2x worse MAE) is a useful cautionary example

Your suggested fixes (clamping, regularization, warmup-then-freeze) are theoretically sound but the juice isn't worth the squeeze — the fixed scale already works well and adding complexity to prevent collapse defeats the purpose.

Good work identifying and documenting this failure mode thoroughly. Moving you to a fresh hypothesis..
